### PR TITLE
Add deepPickProperties function

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
This PR adds a `deepPickProperties` function, which can be used to pull desired properties from an object by providing a list of keys. The implementation supports top-level keys, as well as nested keys by specifying 'nested.properties.like.this'. While the types may support other Grammar, this is not something the JS implementation supports quite yet.

Eventually, the hope is to use this within the service layer, to allow for partial querying of entities to reduce over-the-wire cost.

## 📦 Impacts:
See changesets

### CODE REVIEW

**Comment by avatarneil** [comment]
I've also got an issue out to the authors of the `DeepPick` type library to support grammar restrictions, so the typing can be a little more strict https://gitlab.com/soul-codes/ts-deep-pick/-/issues/7

File: packages/mds-utils/deep-pick-properties.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * ```typescript
+ * // You can pass in no paths in order for deepPickProperties to act as the identity function.
+ * const y: { telemetry: { gps: { lat: 1, lng: 2 } }, event_type: 'foo' } = deepPickProperties({ telemetry: { gps: { lat: 1, lng: 2 } }, event_type: 'foo' })
+ */
+export const deepPickProperties = <T, P extends DeepPickPath<T>[] = ['*']>(
+  entity: T,
+  unparsedPaths?: P
+): P extends ['*'] ? DeepPick<T, '*'> : DeepPick<T, P[number]> => {
+  if (typeof unparsedPaths === 'undefined' || unparsedPaths.length === 0)
+    return entity as P extends ['*'] ? DeepPick<T, '*'> : DeepPick<T, P[number]> // this is a pretty wacky assertion, but is necessary because conditional returns on functions in TS are weird
```
Comment (drtyh2o) [inline packages/mds-utils/deep-pick-properties.ts:27]: Use block syntax {} for the if statement?

File: packages/mds-utils/deep-pick-properties.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * @example
+ * ```typescript
+ * // You can pass in no paths in order for deepPickProperties to act as the identity function.
+ * const y: { telemetry: { gps: { lat: 1, lng: 2 } }, event_type: 'foo' } = deepPickProperties({ telemetry: { gps: { lat: 1, lng: 2 } }, event_type: 'foo' })
+ */
+export const deepPickProperties = <T, P extends DeepPickPath<T>[] = ['*']>(
+  entity: T,
+  unparsedPaths?: P
+): P extends ['*'] ? DeepPick<T, '*'> : DeepPick<T, P[number]> => {
+  if (typeof unparsedPaths === 'undefined' || unparsedPaths.length === 0)
```
Comment (drtyh2o) [inline packages/mds-utils/deep-pick-properties.ts:26]: Could this be `if (!unusedPath?.length) {}`?

File: packages/mds-utils/deep-pick-properties.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    return entity as P extends ['*'] ? DeepPick<T, '*'> : DeepPick<T, P[number]> // this is a pretty wacky assertion, but is necessary because conditional returns on functions in TS are weird
+
+  const paths = unparsedPaths.map(p => p.split('.'))
+
+  return Object.entries(entity).reduce((acc, [key, value]) => {
+    if (paths && paths.some(path => path[0] === key)) {
```
Comment (drtyh2o) [inline packages/mds-utils/deep-pick-properties.ts:32]: Will `paths` ever be undefined here? Will `path` always have at least 1 element to make `path[0]` always safe?

**Review by drtyh2o — APPROVED** [review]
Looks good. Will need to use it in practice to see how helpful it is. Conceptually it makes sense.

File: packages/mds-utils/deep-pick-properties.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-utils/deep-pick-properties.ts:26]: Ahh clever, yeah I think that should work!

File: packages/mds-utils/deep-pick-properties.ts (hunk 3 above)
Comment (avatarneil) [reply packages/mds-utils/deep-pick-properties.ts:32]: I think it'd be possible to break that -- fix incoming

Comment (avatarneil) [reply packages/mds-utils/deep-pick-properties.ts]: Fixed in a138965

File: packages/mds-utils/deep-pick-properties.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-utils/deep-pick-properties.ts]: Fix in a138965

File: packages/mds-utils/deep-pick-properties.ts (hunk 1 above)
Comment (avatarneil) [reply packages/mds-utils/deep-pick-properties.ts:27]: Fixed in a138965

**Review by drtyh2o — APPROVED** [review]

**Review by janedotx — APPROVED** [review]

**Pull request merged**